### HW 1

1) Version of pandas

In [31]:
import numpy as np
import pandas as pd


print(pd.__version__)

2.2.3


In [3]:
import wget

url = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv'
output_directory = './data/laptops.csv'
wget.download(url, out=output_directory)

'./data/laptops.csv'

#### EDA

2) Number of records

In [4]:
df = pd.read_csv('./data/laptops.csv')
df.shape

(2160, 12)

In [7]:
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [6]:
num_of_records = df.shape[0]
print(f"Number of records in laptops.csv: {num_of_records}")

Number of records in laptops.csv: 2160


3) Number of Laptop brands

In [8]:
num_of_laptop_brands = df.Brand.nunique()
print(f"Number of laptop brands in the dataset: {num_of_laptop_brands}")

Number of laptop brands in the dataset: 27


4. Number of columns in the dataset with missing values

In [10]:
num_cols_missing_values = df.isna().sum()
num_cols_missing_values

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

In [16]:
cols_missing_values = [col for col in df.columns if df[col].isna().any()]
print(f"Columns with missing values: {cols_missing_values}")
print(f"Number of columns with missing values {len(cols_missing_values)}")

Columns with missing values: ['Storage type', 'GPU', 'Screen']
Number of columns with missing values 3


5) Maximum final price

In [25]:
dell_max = df[df['Brand']=='Dell']['Final Price'].max().item()
print(f"Maximum final price of Dell notebooks in the dataset is: ${dell_max}")

Maximum final price of Dell notebooks in the dataset is: $3936.0


6. Median value of screen

In [29]:
print(f"Median without fillna: {df.Screen.median()}")
print(f"Mode {df.Screen.mode()}")
df['Screen'] = df['Screen'].fillna(df.Screen.mode())
print(f"Median value after fill is: {df.Screen.median()}")

Median without fillna: 15.6
Mode 0    15.6
Name: Screen, dtype: float64
Median value after fill is: 15.6


No change to the median value

7. Sum of weights

In [30]:
df_innjoo = df[df['Brand']=='Innjoo']
df_innjoo.head()

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02

In [44]:
X = df_innjoo[['RAM','Storage','Screen']].values
XTX = X.T @ X
XTX_inv = np.linalg.inv(XTX)
y = np.array([1100, 1300, 800,900,1000,1100])
w = XTX_inv @ X.T @ y
w = np.sum(w)
w


np.float64(91.2998806299555)

Sum of all the elements of the result is 91.30